vpr/src/place/analytical_placement/AnalyticalSolver.cpp

```diff
@@ -162,21 +211,255 @@ void QPHybridSolver::solve(unsigned iteration, PartialPlacement &p_placement) {
+    }
 }
 
-bool QPHybridSolver::isSemiPosDef(const Eigen::SparseMatrix<double> &A){
-    // TODO: This is slow, it could be faster if we use Cholesky decomposition (Eigen::SimplicialLDLT), still O(n^3)
-    Eigen::SelfAdjointEigenSolver<Eigen::SparseMatrix<double>> solver;
-    solver.compute(A);
-    VTR_ASSERT(solver.info() == Eigen::Success && "Eigen solver failed!");
-    Eigen::VectorXcd eigenvalues = solver.eigenvalues();
-    for (int i = 0; i < eigenvalues.size(); ++i) {
-        VTR_ASSERT(eigenvalues(i).imag() == 0 && "Eigenvalue not real!");
-        if(eigenvalues(i).real()<=0){
-            return false;
+void B2BSolver::initialize_placement_random_uniform(PartialPlacement &p_placement) {
+    size_t grid_width = g_vpr_ctx.device().grid.width();
+    size_t grid_height = g_vpr_ctx.device().grid.height();
+    // std::random_device rd;
+    // std::mt19937 gen(rd());
+    std::mt19937 gen(1894650209824387);
+    std::uniform_real_distribution<> disx(0.0, std::nextafter(grid_width, 0.0));
+    std::uniform_real_distribution<> disy(0.0, std::nextafter(grid_height, 0.0));
+    for (size_t i = 0; i < p_placement.num_moveable_nodes; i++) {
+        p_placement.node_loc_x[i] = disx(gen);
+        p_placement.node_loc_y[i] = disy(gen);
+    }
+}
+
+void B2BSolver::initialize_placement_least_dense(PartialPlacement &p_placement) {
+    size_t grid_width = g_vpr_ctx.device().grid.width();
+    size_t grid_height = g_vpr_ctx.device().grid.height();
+    unsigned cols = std::ceil(std::sqrt(p_placement.num_moveable_nodes * grid_height / static_cast<double>(grid_width)));
+    unsigned rows = std::ceil(p_placement.num_moveable_nodes / static_cast<double>(cols));
+    double x_gap = grid_width / static_cast<double>(cols + 1);
+    double y_gap = grid_height / static_cast<double>(rows + 1);
+    for (int r = 1; r <= static_cast<int>(rows); r++) {
+        for (int c = 1; c <= static_cast<int>(cols); c++) {
+            p_placement.node_loc_x[(r-1)*cols + c -1] = c * x_gap;
+            p_placement.node_loc_x[(r-1)*cols + c -1] = r * y_gap;
         }
     }
-    return true;
+}
+
+// This function return the two nodes on the bound of a netlist, (max, min)
+std::pair<size_t, size_t> B2BSolver::boundNode(std::vector<size_t>& node_ids, std::vector<double>& node_locs){
+    auto compare = [&node_locs](int a, int b) {
+        return node_locs[a] < node_locs[b];
+    };
+    auto maxIt = std::max_element(node_ids.begin(), node_ids.end(), compare);
+    auto minIt = std::min_element(node_ids.begin(), node_ids.end(), compare);
+    VTR_ASSERT(node_locs[*maxIt] >= node_locs[*minIt] && "max note should be greater than min node");
+    return std::make_pair(*maxIt, *minIt);
+}
+
+void B2BSolver::populate_matrix(PartialPlacement &p_placement) {
+    const AtomContext& atom_ctx = g_vpr_ctx.atom();
+    const AtomNetlist& netlist = p_placement.atom_netlist;
+    // Resetting As bs
+    A_sparse_x.setZero();
+    A_sparse_y.setZero();
```

Comment:
I think eigen only records non zero terms for a sparse matrix. When set zero, all the information within the sparse matrix data structures are memseted to represent no entry, but not resized. It could help with reallocation, but of course performence is not the main focus. I will do A_sparse = SparseMatrix<double>...

Reply:
I think thats a good idea. We want to prevent early optimization. In the future if we find that allocations are causing slow-down we can revisit this; but we want to avoid any unexpected bugs from creeping in!